File: app/src/main/java/com/example/kkocel/marvel/ComicDetailActivity.java
package com.example.kkocel.marvel;

import android.app.ListActivity;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.app.ActionBar;
import android.view.MenuItem;

public class ComicDetailActivity extends AppCompatActivity {

    public static final String COMIC_ID ="COMIC_ID";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_comic_detail);
        Toolbar toolbar = (Toolbar) findViewById(R.id.detail_toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        if (savedInstanceState == null) {
            Bundle arguments = new Bundle();
            arguments.putString(ComicDetailFragment.ARG_ITEM_ID,
                    getIntent().getStringExtra(ComicDetailFragment.ARG_ITEM_ID));
            ComicDetailFragment fragment = new ComicDetailFragment();
            fragment.setArguments(arguments);
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.comic_detail_container, fragment)
                    .commit();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            navigateUpTo(new Intent(this, ListActivity.class));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/kkocel/marvel/detail/view/ComicDetailActivity.kt
package com.example.kkocel.marvel.detail.view

import android.app.ListActivity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.view.animation.AlphaAnimation
import com.bumptech.glide.Glide
import com.example.kkocel.marvel.R
import com.example.kkocel.marvel.detail.di.DetailModule
import com.example.kkocel.marvel.detail.presenter.DetailPresenter
import com.example.kkocel.marvel.detail.state.ComicDetailViewState
import com.example.kkocel.marvel.network.rest.RetrofitModule
import com.example.kkocel.marvel.state.ConnectionErrorViewState
import com.example.kkocel.marvel.view.ThumbnailUrlLoader
import com.example.kkocel.marvel.view.displaySnackbar
import jp.wasabeef.glide.transformations.BlurTransformation
import kotlinx.android.synthetic.main.activity_comic_detail.*


class ComicDetailActivity : AppCompatActivity(), DetailView {

    companion object {
        val COMIC_ID = "COMIC_ID"
    }

    private val INVALID_COMIC_ID = -1

    private lateinit var detailPresenter: DetailPresenter
    private var selectedComic: Int = INVALID_COMIC_ID

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        selectedComic = intent.getIntExtra(COMIC_ID, INVALID_COMIC_ID)

        if (selectedComic == INVALID_COMIC_ID) {
            throw IllegalStateException("Can't run detail activity without ID")
        }

        val detailModule = DetailModule(RetrofitModule())

        setContentView(R.layout.activity_comic_detail)
        setupToolbar()

        detailPresenter = detailModule.provideListPresenter(this)
        detailPresenter.onCreate(selectedComic)

        swpDetailRefreshLayout.isRefreshing = true
        swpDetailRefreshLayout.setOnRefreshListener {
            detailPresenter.forceReload(selectedComic)
        }
    }

    private fun setupToolbar() {
        setSupportActionBar(tlbDetailToolbar)
        tlbDetailToolbar.title = title
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    fun fadeIn(view: View) {
        val animate = AlphaAnimation(0f, 1f)
        animate.duration = 500
        animate.fillAfter = true
        view.startAnimation(animate)
    }

    override fun onDetailsLoaded(detailViewState: ComicDetailViewState) {
        crdDetailContent.visibility = View.VISIBLE
        fadeIn(crdDetailContent)
        swpDetailRefreshLayout.isRefreshing = false
        txtDetailComicName.text = detailViewState.title
        txtDetailComicOverview.text = detailViewState.description
        Glide.with(this).using(ThumbnailUrlLoader(this)).load(detailViewState.thumbnail).into(imgDetailThumbnail)
        Glide.with(this).using(BackdropDetailUrlLoader(this)).load(detailViewState.image)
                .bitmapTransform(BlurTransformation(this, 10))
                .into(imgDetailBackdrop)
    }

    override fun onNetworkError(detailViewState: ConnectionErrorViewState) {
        displaySnackbar(getString(R.string.noNetwork))
        detailPresenter.retryWhenNetworkIsAvailable(selectedComic)
    }

    override fun onUnrecoverableError(e: Throwable) {
        displaySnackbar(getString(R.string.seriousProblem))
        // TODO: Log to Crashlytics?
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            navigateUpTo(Intent(this, ListActivity::class.java))
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        detailPresenter.onViewDestroyed()
    }
}
